handler/src/test/java/io/netty/handler/ssl/SslHandlerTest.java

```diff
@@ -431,6 +437,97 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exc
+                .childHandler(new ChannelInitializer<Channel>() {
+                  @Override
+                  protected void initChannel(Channel ch) {
+                      ch.pipeline().addLast(sslServerCtx.newHandler(ch.alloc()));
+                      ch.pipeline().addLast(new ChannelInboundHandlerAdapter() {
+                          @Override
+                          public void channelActive(ChannelHandlerContext ctx) {
+                              ByteBuf buf = ctx.alloc().buffer(10);
+                              buf.writerIndex(buf.readerIndex() + 10);
+                              ctx.writeAndFlush(buf).addListener(new ChannelFutureListener() {
+                                  @Override
+                                  public void operationComplete(ChannelFuture future) {
+                                      events.add(future);
+                                      latch.countDown();
+                                  }
+                              });
+                          }
+
+                          @Override
+                          public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
+                              if (evt instanceof SslCompletionEvent) {
+                                  events.add(evt);
+                                  latch.countDown();
+                                  latch2.countDown();
+                              }
+                          }
+                      });
+                  }
+                });
+
+            Bootstrap cb = new Bootstrap();
+            cb.group(group)
+                .channel(LocalChannel.class)
+                .handler(new ChannelInitializer<Channel>() {
+                    @Override
+                    protected void initChannel(Channel ch) {
+                        ch.pipeline().addFirst(new ChannelInboundHandlerAdapter() {
+                            @Override
+                            public void channelActive(ChannelHandlerContext ctx) {
+                                ByteBuf buf = ctx.alloc().buffer(1000);
+                                buf.writerIndex(buf.readerIndex() + 1000);
+                                ctx.writeAndFlush(buf);
+                            }
+                        });
+                    }
+                });
+
+            serverChannel = sb.bind(new LocalAddress("SslHandlerTest")).sync().channel();
+            clientChannel = cb.connect(serverChannel.localAddress()).sync().channel();
+            latch.await();
+
+            SslCompletionEvent evt = (SslCompletionEvent) events.take();
+            assertTrue(evt instanceof SslHandshakeCompletionEvent);
+            assertThat(evt.cause(), is(instanceOf(SSLException.class)));
+
+            ChannelFuture future = (ChannelFuture) events.take();
+            assertThat(future.cause(), is(instanceOf(SSLException.class)));
+
+            serverChannel.close();
+            clientChannel.close();
```

Comment:
@Scottmitch nit: also null these out so we not try to close again in the finally block.

Reply:
Also you may want to call `sync()`, not needed but still not bad I think

Reply:
done